### ISSUE TITLE
Question about Primitive Loss

### ISSUE THREAD

**Issue by haoala** [kimren227/CSGStumpNet#6]
Hi, thank you for this interesting work and congrats on publishing at ICCV.

I have a question about the implementation of the primitive loss term. In the paper, it is

<img src="https://render.githubusercontent.com/render/math?math=L_{primitive} = \frac{1}{K} \sum_k \min SDF_k^2 (x_n)">

But in the implementation it is:
https://github.com/kimren227/CSGStumpNet/blob/e23c6fbd20097000b886221abd70def34ac2a15b/loss.py#L11

Which seems like it is this expression instead:

<img src="https://render.githubusercontent.com/render/math?math=L_{primitive} = \frac{1}{K} \sum_k (\min SDF_k (x_n)) ^ 2">

I suppose this is equivalent to the first equation when all testing points are outside the primitive, but once there are testing points inside this loss term seems to pull the surface so that no testing points are too far inside (since it pushes the most negative value toward 0). Is this intentional?

Might the correct implementation be as follows?
```py
primitive_loss = torch.mean((primitive_sdf**2).min(dim=1)[0]) * self.scale
```

**Comment by kimren227** [comment]
Hi Adriel,

Thank you for going through the code in detail and I really appreciate that.

This is a bug. And I believe fixing it can increase the performance of the model.

Do you want to make a PR for this?

Best Regards,

Daxuan

**Comment by haoala** [comment]
Hi Daxuan,

If this new loss function is untested, and the model that was trained on the existing loss function works fine, then I would be more leaving it as is until it is tested, in case of unintended side effects. So I am happy to leave this as an open issue.

**Comment by Farewell-ME** [comment]
Hi, I'd like to ask about figure 1 and 3 in your paper. Because these 3 demo in two figures are all design tree for mechanical parts, which may meet the demand of my research interest. However,  the training data in your code is modified from ShapeNet, which is composed of complex assemblies. 
I myself have made several types of mechanical parts point cloud to train CSGstumpNet. But the performance is not good. Could you kindly publish the mechnical parts point cloud data used for Figure 1 and 3?

**Comment by kimren227** [comment]
Hi Frank,

Thanks for your interest. 

Figure 1 and 3 are not trained using the end2end pipeline(the code in this repo). Instead, they are used for the optimization method as described in Sec.5.1. Currently, I'm hesitant to release the code for the optimization method as it was implemented in a hurry and the code is quite messy. But if you really want to give it a shot, please PM me, I can send you a copy of the code.

Best Regards.

Daxuan

**Comment by Farewell-ME** [comment]
Oh, thank you so much! I'm new to Github and it seems not to have a PM button. Is the education email address left on your paper OK to contact you?

**Comment by kimren227** [comment]
Yes, that will work

**Comment by kimren227** [comment]
I'll email you with the code. Do let me know if you need further assistance.

**Comment by rucha97** [comment]
Hi Daxuan
I found your work quite interesting. Kudos!

I would like to know how I should visualize the results on OpenSCAD, I have never worked on that. Perhaps if you had any steps, that would be quite helpful. TIA!

Cheers,
Rucha

**Comment by kimren227** [comment]
> 

You may follow #11 to export the network result into an OpenSCAD file. After that, you can directly open it using the software.

**Comment by rucha97** [comment]
> > 
> 
> You may follow #11 to export the network result into an OpenSCAD file. After that, you can directly open it using the software.

Oh I see, I did not check the closed issues. Thank you!
